storagenode/contact/chore.go

```diff
@@ -48,53 +58,108 @@ func (chore *Chore) Run(ctx context.Context) (err error) {
 	defer mon.Task()(&ctx)(&err)
 	chore.log.Info("Storagenode contact chore starting up")
 
+	var group errgroup.Group
+
 	if !chore.service.initialized.Wait(ctx) {
 		return ctx.Err()
 	}
+	chore.mu.Lock()
+	for _, satellite := range chore.trust.GetSatellites(ctx) {
+		satellite := satellite
+		backOff := time.Second
+		interval := chore.interval
+		cycle := sync2.NewCycle(interval)
+		chore.cycles = append(chore.cycles, cycle)
+
+		cycle.Start(ctx, &group, func(ctx context.Context) error {
+			chore.log.Info("starting contact cycle for satellite " + satellite.String())
+			err := chore.pingSatellite(ctx, satellite)
+			if err == nil {
+				return nil
+			}
 
-	return chore.Loop.Run(ctx, func(ctx context.Context) error {
-		if err := chore.pingSatellites(ctx); err != nil {
-			chore.log.Error("pingSatellites failed", zap.Error(err))
-		}
-		return nil
-	})
+			interval := backOff
+			retries := 0
+			for {
+				if !sync2.Sleep(ctx, interval) {
+					chore.log.Error("ping satellite failed after " + strconv.Itoa(retries) + " retries timed out")
+					return ctx.Err()
+				}
+				err := chore.pingSatellite(ctx, satellite)
+				retries++
+				if err == nil {
+					return nil
+				}
+				chore.log.Error("ping satellite failed " + strconv.Itoa(retries) + " times")
+				interval *= 2
+				if interval > chore.interval {
+					interval = chore.interval
+				}
+			}
+		})
+	}
+	chore.mu.Unlock()
+	chore.started.Release()
+	return group.Wait()
 }
 
-func (chore *Chore) pingSatellites(ctx context.Context) (err error) {
+func (chore *Chore) pingSatellite(ctx context.Context, id storj.NodeID) (err error) {
 	defer mon.Task()(&ctx)(&err)
-	var group errgroup.Group
 	self := chore.service.Local()
-	satellites := chore.trust.GetSatellites(ctx)
-	for _, satellite := range satellites {
-		satellite := satellite
-		addr, err := chore.trust.GetAddress(ctx, satellite)
-		if err != nil {
-			chore.log.Error("getting satellite address", zap.Error(err))
-			continue
-		}
-		group.Go(func() error {
-			conn, err := chore.dialer.DialAddressID(ctx, addr, satellite)
-			if err != nil {
-				return err
-			}
-			defer func() { err = errs.Combine(err, conn.Close()) }()
+	address, err := chore.trust.GetAddress(ctx, id)
+	if err != nil {
+		return errPingSatellite.New("failed to get satellite address %s: %v", id.String(), err)
```

Comment:
not sure if this should be a `pingSatellite` error, since we haven't attempted to ping the satellite yet

Reply:
on a second look, I think it's ok since it's the error class for the ping satellite method